src/main/java/com/google/api/codegen/discovery/DefaultString.java

```diff
@@ -28,12 +33,76 @@
  */
 public class DefaultString {
 
+  @AutoValue
+  abstract static class SampleKey {
+    abstract String getApiName();
+
+    abstract String getFieldName();
+
+    abstract String getRegexp();
+
+    static SampleKey create(String apiName, String fieldName, String regexp) {
+      return new AutoValue_DefaultString_SampleKey(apiName, fieldName, regexp);
+    }
+  }
+
+  private final String declare;
+  private final String comment;
+
+  private DefaultString(String declare, String comment) {
+    this.declare = declare;
+    this.comment = comment;
+  }
+
+  public String getDeclare() {
+    return declare;
+  }
+
+  public String getComment() {
+    return comment;
+  }
+
+  private static final ImmutableMap<SampleKey, String> SAMPLE_STRINGS =
+      ImmutableMap.<SampleKey, String>builder()
+          .put(
+              SampleKey.create("compute", "zone", "[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?"),
+              "us-central1-f")
+          .put(
+              SampleKey.create("autoscaler", "zone", "[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?"),
+              "us-central1-f")
+          .put(
+              SampleKey.create("clouduseraccounts", "zone", "[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?"),
+              "us-central1-f")
+          .build();
+
+  public static DefaultString of(String apiName, String fieldName, String pattern) {
+    String sample = null;
+
+    if (pattern != null) {
+      // If the pattern has a specially-recognized default, use the default. No sample.
+      String def = forPattern(pattern);
+      if (def != null) {
+        return new DefaultString(def, null);
+      }
+
+      // If the pattern has a specially-recognized sample, use the sample.
+      sample = SAMPLE_STRINGS.get(SampleKey.create(apiName, fieldName, pattern));
+    }
+
+    String def =
+        String.format(
+            "{MY-%s}", LanguageUtil.lowerCamelToUpperUnderscore(fieldName).replace('_', '-'));
+
+    return new DefaultString(def, sample);
+  }
+
   private static final String WILDCARD_PATTERN = "[^/]*";
 
   /**
    * Returns a default string from `pattern`, or null if the pattern is not supported.
    */
-  public static String forPattern(String pattern) {
+  @VisibleForTesting
+  static String forPattern(String pattern) {
     // We only care about patterns that has alternating literal and wildcard like
```

Comment:
Fix the has/have.

You're right. A more rigorous check was added in `validElems` but I didn't delete the old. The old check is deleted now.